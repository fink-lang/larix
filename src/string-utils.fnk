unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
