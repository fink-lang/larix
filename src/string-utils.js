import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

