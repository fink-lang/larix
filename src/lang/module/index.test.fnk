{describe, it, expect, to_match_snapshot} = import '@fink/jest'

{parse} = import '../../'


describe 'module', fn:

  it 'parses module', fn:
    expect
      parse '
        # leading module comment

        # expr comment
        foo = bar
        spam = ni

        # trailing module comment
      '
      to_match_snapshot
