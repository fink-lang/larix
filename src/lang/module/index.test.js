import {strip_block} from '../../string-utils';
import {parse, parse_expr} from '../../';


describe('module', ()=> {

  it('parses module', ()=> {
    const mod = parse(strip_block`
      foo = bar
      spam = ni
    `);

    expect(mod).toMatchSnapshot();
  });
});


describe('module parse errors', ()=> {

  it('throws with bad indentation', ()=> {
    expect(()=> parse(strip_block`
      match fooobar:
        match spam:
        shrub`, 'test.fnk')
    ).toThrow(strip_block`
      test.fnk:3:2
      1| match fooobar:
      2|   match spam:
      3|   shrub
           ^

      Expected indentation > 2.
      `
    );
  });

});
