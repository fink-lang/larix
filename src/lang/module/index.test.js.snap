// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module parses module 1`] = `
module (1:0-2:9)
  assign = (1:0-1:9)
    other (1:0-1:3) foo
    other (1:6-1:9) bar
  assign = (2:0-2:9)
    other (2:0-2:4) spam
    other (2:7-2:9) ni
`;
